kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/usb/UsbModemDevice.java

```diff
@@ -21,80 +21,133 @@
 
     /** The TTY devices associated with modem **/
-    private final ArrayList<String> m_ttyDevs;
+    private final ArrayList<TtyDev> ttyDevs;
 
     /** The block devices associated with the modem **/
-    private final ArrayList<String> m_blockDevs;
+    private final ArrayList<String> blockDevs;
 
     public UsbModemDevice(String vendorId, String productId, String manufacturerName, String productName,
             String usbBusNumber, String usbDevicePath) {
         super(vendorId, productId, manufacturerName, productName, usbBusNumber, usbDevicePath);
 
-        this.m_ttyDevs = new ArrayList<String>();
-        this.m_blockDevs = new ArrayList<String>();
+        this.ttyDevs = new ArrayList<>();
+        this.blockDevs = new ArrayList<>();
     }
 
     public UsbModemDevice(AbstractUsbDevice usbDevice) {
         super(usbDevice);
 
-        this.m_ttyDevs = new ArrayList<String>();
-        this.m_blockDevs = new ArrayList<String>();
+        this.ttyDevs = new ArrayList<>();
+        this.blockDevs = new ArrayList<>();
     }
 
     @Override
     public List<String> getSerialPorts() {
-        return Collections.unmodifiableList(this.m_ttyDevs);
+        List<String> serialPorts = new ArrayList<>();
+        for (TtyDev dev : this.ttyDevs) {
+            serialPorts.add(dev.getPortName());
+        }
+        return serialPorts;
     }
 
     /**
-     * @return sorted list of tty devs
+     * Return a list of tty devices, sorted in a way to facilitate the client code identifying dedicated devices,
+     * e.g. for AT commands, PPP link or NMEA sentences, based on their position in the list.
+     * Originally, only the tty name was used for the comparison. This proved to be wrong as a tty name does reliably
+     * identify the USB interface number (bInterfaceNumber) of the tty device.
+     * To preserve the API contract, the tty devices can be added specifying the USB interface number and this will be
+     * used to sort the list.
+     *
+     * @return sorted list of tty devices
      */
     public List<String> getTtyDevs() {
-        return Collections.unmodifiableList(this.m_ttyDevs);
+        return getSerialPorts();
     }
 
     /**
-     * @return sorted list of block devs
+     * @return sorted list of block devices
      */
     public List<String> getBlockDevs() {
-        return Collections.unmodifiableList(this.m_blockDevs);
+        return Collections.unmodifiableList(this.blockDevs);
+    }
+
+    /**
+     * Adds a tty device identified by its name and USB interface number (bInterfaceNumber).
+     * The devices will be sorted by the interface number. If this is missing, the name will be used.
+     *
+     * @param ttyDev
+     *            the name of the tty device
+     * @param interfaceNumber
+     *            the number of the interface as described by the bInterfaceNumber property
+     */
+    public void addTtyDev(String ttyDev, Integer interfaceNumber) {
```

Comment:
I believe that we agree that interfaceNumber **should not be null**. Otherwise `addTtyDev("foo", null)` would be equivalent to `addTtyDev("foo")` which is now deprecated.
Forcing the interface number to be non null would be the right decision but, paradoxically, the current implementation can recover such a broken modem if we let the interface number to be null.